src/tagged-type.ts

```diff
@@ -0,0 +1,117 @@
+export type Module = string;
+export type DocumentReference = string;
+
+/**
+ * TaggedType provides the encoding/decoding of the Fauna Tagged Type formatting
+ */
+export class TaggedTypeFormat {
+  /**
+   * Encode the Object to the Tagged Type format for Fauna
+   *
+   * @param obj - Object that will be encoded
+   * @returns Map of result
+   */
+  static encode(obj: any): any {
+    return new taggedTypeEncoded(obj).result;
+  }
+
+  /**
+   * Decode the JSON string result from Fauna to remove Tagged Type formatting.
+   *
+   * @param input - JSON string result from Fauna
+   * @returns object of result of FQL query
+   */
+  static decode(input: string): any {
+    return JSON.parse(input, (_, value: any) => {
+      if (value["@mod"]) {
+        return value["@mod"] as Module;
+      } else if (value["@doc"]) {
+        return value["@doc"] as DocumentReference;
+      } else if (value["@int"]) {
+        return Number(value["@int"]);
+      } else if (value["@long"]) {
+        return Number(value["@long"]);
+      } else if (value["@decimal"]) {
+        return Number(value["@decimal"]);
+      } else if (value["@double"]) {
+        return Number(value["@double"]);
```

Comment:
Is it a problem that all of these values lose their types by the time you send them back to Fauna? Do we rely on type constructors in fqlx when stored type is important, e.g. `Int(${some_number})`?

This is not a concern when we have schema enforcement, so maybe that's the real answer here.

Javascript doesn't allow you to override operators, so the closest thing I think we could provide would be a class that implements `valueOf` so JS can coerce the value when it is used as a number. But then we are back to not having a type if you try to plug the result back into a query.

Reply:
Yeah I'm not quite sure how to handle this with JavaScript since it only has the one type. Maybe we need to provide our own types like we're doing for `Module` and `DocumentReference`.

Reply:
I mostly wanted to spark the conversation and thoughts from others. It is reasonable to me to move forward decoding to `Number` at this time.

Reply:
> Maybe we need to provide our own types like we're doing for Module and DocumentReference.

I see this convo is resolved, but this is what popped into my mind, as well. Did you discuss elsewhere?

Reply:
No just resolved to move forward with this PR.

If we made our own JS classes for things like `FaunaInt`, `FaunaDecimal` and even `FaunaDate` and `FaunaTime` (the latter of which we have in the v4 driver), then I don't know how useful it would be to _decode_ responses into these classes. 

As I was working later on testing, I realized how persnickety it is to encode `Date` vs `Time` types. i.e. a JS `Date` is only a Fauna `Date` if it's at 00:00:00 UTC. It would be nice for users to cast use a JS `Date` as a Fauna `Date` regardless of the time.

That makes me think back to numbers. We've had users want to store floating point values that happen to be whole numbers. So it would be nice to have a mechanism to explictly cast numbers as either Fauna ints or floats.

Now, folks can cast things now using FQL constructors, but that is limited to the parts that you handle directly in FQL. Sometimes you have an object in code you just want to store, e.g. `somedoc.update(${data})` where data contains fields that should be cast. Passing around explicit Fauna types could help with that.

I would support a fast follow up to add date and time classes and decode responses to those instead of immediately to Date.

If would also support classes for int, long, decimal & double. However, I am not convinced we should decode directly to those, but rather we could have them as a mechanism for users to ensure their values get encoded in a consistent way.

Reply:
We should probably make a decision on this. Post scrum topic today!

Reply:
We should probably make a decision on this. Post scrum topic today!

Reply:
> If would also support classes for int, long, decimal & double. However, I am not convinced we should decode directly to those, but rather we could have them as a mechanism for users to ensure their values get encoded in a consistent way.

regarding this, the problem is that we need to guarantee safe round-tripping. we'll have to reach consensus, but i think of this as a hard requirement

Reply:
Number handling is updated using `number` and `BigInt` without any custom classes. I've stubbed out classes and notes for `FaunaDate` and `FaunaTime`. We will model after the TC39 proposed `Temporal.PlainDate` and `Temporal.Instant`, respectively, with the intent to replace these implementations with the Temporal types when they are stable.